Complete overhaul of the extension

dhruvkb opened the pull request:
Problem:
Nothing used to work due to missing APIs on Channel-i, the code was inconsistent at best and completely non-compliant with JavaScript code style guidelines at worst.

Solution:
New code is consistent and readable and reduced in bulk too.
README.md has been cleaned, punctuated and updated with the new resources.
Style has been modernised.
Everything just works.

Please merge this! I want to be a member of the organization :smile:

dhruvkb committed:
Make the HTML more semantic (pun intentional)

dhruvkb committed:
Use the latest version of jQuery served from the jQuery website itself

dhruvkb committed:
Remove all the old CSS to make way for the modern CSS

dhruvkb committed:
Remove all the out-of-fashion fonts to make way for modern fonts

dhruvkb committed:
Replace the tiny IMG logo with a larger one

dhruvkb committed:
Format the code, copy the description from popup.html and remove the background listing of jquery-1.9.1.js

dhruvkb committed:
Remove the forward slash from /> in the unpaired <img> tag

dhruvkb committed:
Update the loading spinner to be half its size for Retina support and give it a better name

dhruvkb committed:
Complete revamp of the JavaScript, most notably the switch from the old API to the LecTut API and the better code consistency

dhruvkb committed:
Fix a bug: HOST written as host, which would lead to problems

dhruvkb committed:
Fix a bug: DOMAIN written as domain, which would lead to problems

dhruvkb committed:
Add the 'Login' button and move the JS files from the Internet to local as Chrome does not allow extensions to load external JS files

dhruvkb committed:
Add the IMG logo and remove the animation from the button

dhruvkb committed:
Style the logo and center all the content

dhruvkb committed:
Manage the showing and hiding of different views

dhruvkb committed:
Create a user region and replace all the _ in id attributes with -

dhruvkb committed:
Load the user information in the respective fields and replace all the _ in id attributes with -

dhruvkb committed:
Add styles for the user photo and replace all the _ in id attributes with -

dhruvkb committed:
Increase the width and consequently shrink the size of user-photo

dhruvkb committed:
Add a logout button and create a table of application links

dhruvkb committed:
Add jQuery to the background scripts and http://info.cern.ch/* to the permissions so that it can be contacted to check for connectivity

dhruvkb committed:
Fix code formatting, document functions and correct the code to use the LecTut API instead of the defunct old API

dhruvkb committed:
Add code to change the icons depending on whether the user is logged in or not

dhruvkb committed:
Add the credits on the main page

dhruvkb committed:
Make the IMG logo a link to the Facebook Page of IMG

dhruvkb committed:
Fix a missing - in the documentation

dhruvkb committed:
Restructure the code to make it modular, which would make it easy to implement the 'sync login/logout state across tabs' feature, in case it is decided to implement it

dhruvkb committed:
Give the popup the background color the same as the color of the chrome of Google Chrome

dhruvkb committed:
Replace the basic buttons with labeled icon buttons

dhruvkb committed:
Fix a few missing spaces in the license text

dhruvkb committed:
Add new screenshots after the redesign

dhruvkb committed:
Make massive changes to the README as the final step and add my name to the list of contributors to the project

ritz078 commented on css/popup.css:
```diff
@@ -1,177 +1,57 @@
+.centered-content {
+    height: 100%;
+
+    display: flex;
```
👍

dhruvkb replied:
:smiley:

ritz078 commented on js/apps.js:
```diff
@@ -0,0 +1,78 @@
+
+var DOMAIN = "https://channeli.in/";
+
+buysell = new App("Buy & Sell", "buysell", "buyandsell");
```
you are using global declarations. Will throw error when google implements 'use-strict' by default. Better change this to `var buysell`

dhruvkb replied:
Fixed

ritz078 commented on js/apps.js:
```diff
@@ -0,0 +1,78 @@
+    shp,
+    thinktank
+    //vle
+];
```
why not this ?

``` js
var app = {
  buysell: new App("Buy & Sell", "buysell", "buyandsell"),
  // others
}
```

Accessing this will be also easy by iteration through the keys.

dhruvkb replied:
Do you mean
`var app = {
  "buysell": new App("Buy & Sell", "buysell", "buyandsell"),
  //others
}`

dhruvkb replied:
Fixed

ritz078 commented on js/background.js:
```diff
@@ -1,80 +1,76 @@
+ * @param href - the URL taken from the address bar
+ * @returns {string} hostname - the host name extracted from the URL
+ */
+var getHostName = function (href) {
```
use consistent functions. Somewhere its a function expression and somewhere its function declaration. 

function expression:

``` js
var x = function(){}
```

function declaration

``` js
function x(){}
```

dhruvkb replied:
Careless mistake. Fixed

ritz078 commented on js/background.js:
```diff
@@ -1,80 +1,76 @@
+/** 0 implies 'not logged in' and 1 implies 'logged in' */
+var userStatus = -1;
+/** 0 implies 'not connected' and 1 implies 'connected' */
+var networkStatus = 0;
```
why not true or false ?

dhruvkb replied:
Fixed

ritz078 commented on js/popup.js:
```diff
@@ -1,228 +1,99 @@
+     * @param href - the URL taken from the address bar
+     * @returns {string} hostname - the host name extracted from the URL
+     */
+    var getHostName = function (href) {
```
can be made into a separate file. this function is common in one more file.

dhruvkb replied:
Fixed

ritz078 commented on js/popup.js:
```diff
@@ -1,228 +1,99 @@
+                userStatus = 1;
+                chrome.browserAction.setIcon({path: "../images/icon_active.png"});
+                // Update the popup view
+                var user__username = data.user.username;
```
use camelCase

dhruvkb replied:
Yeah, habit carried over from Python. Fixed

ritz078 commented on popup.html:
```diff
@@ -1,51 +1,75 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+
+    <!-- Semantic UI CSS -->
+    <link rel="stylesheet" type="text/css" href="http://semantic-ui.com/dist/semantic.min.css">
```
It seems like an overkill to use such frameworks for a chrome extension.

dhruvkb replied:
Made the extension look better and took a lot of the designer work off my shoulders. I have a terrible taste in design :smile:

dhruvkb replied:
I removed the JS associated with that framework, will remove all the unused CSS styles as well, that should reduce the burden

ritz078 reviewed (commented):
Hi, I have put the comments for the JS part. I always wanted to remove jQuery. Maybe you can try that. Expect that I guess IMG seniors should take a look. 😄 . I am not sure about showing nothing to the user when he/she is logged out (maybe something has changed in channel-i). Cheers.

jagannath93 commented on js/apps.js:
```diff
@@ -0,0 +1,78 @@
+    shp,
+    thinktank
+    //vle
+];
```
You better fetch the app listing template through ajax from the backend and insert it here in the extension dynamically.
Coz this won't force you to update the extension in the Webstore every time you add or delete an app in the Channeli, everything will be updated automatically

dhruvkb replied:
Working on it

jagannath93 commented on js/apps.js:
```diff
@@ -0,0 +1,78 @@
+        chrome.tabs.create({url: $(this).attr("href")});
+        return false;
+    });
+});
```
The selectable app listing table that you are constructing here in Js, can be removed if you fetch this template from the backend itself.

Note: It is a good practice to not to use HTML inside JS

ritz078 replied:
You are going to hate React then 😄

dhruvkb replied:
The backend doesn't serve this (or anything else for that matter) anymore. Those APIs are gone.

jagannath93 commented on js/background.js:
```diff
@@ -1,80 +1,76 @@
+var checkNetConnection = function () {
+    $.get(NET_CHECK_DOMAIN, {}, function () {
+        if (networkStatus == 0) {
+            checkSession(-1);
```
Why 'checkSession' is accepting an argument '-1' here ? Where is the corresponding function declaration for this call ?

dhruvkb replied:
That got left over accidentally from when I was trying out another logic behind checkSession(). Fixed

jagannath93 commented on js/background.js:
```diff
@@ -1,80 +1,76 @@
 
-var Domain = "https://channeli.in";
-var Host = "channeli.in";
+var NET_CHECK_DOMAIN = "http://info.cern.ch/";
```
Use Channeli's static files serving Nginix endpoint here(You can ask Kartik about the details) instead of some external endpoint, to make this extension to be in sync with channeli.

dhruvkb replied:
This URL returns the fastest response I have ever seen as it is quite possibly the simplest page on the Internet. No CSS/JS, oldest possible HTML and a few lines of plain text only... the only use of this was to check Internet availability

jagannath93 replied:
If the internet goes down, the intranet still works. So the NET_CHECK_DOMAIN should be a channeli url to keep the extension up and running in such instances

dhruvkb replied:
I never thought about that! You're right... added to my to-dos 👍

dhruvkb replied:
Fixed after consultation with @kartikp1995

dhruvkb commented:
@jagannath93 
The problem is the APIs that were used in the extension are no longer present on Channel-i and I'm not quite keen on creating a new set of views for an extension that has limited use. So I used whatever API gave me the three things I needed (the name, enrollment number and photo), which happens to be the LecTut API.

@ritz078 
Why is using jQuery a bad idea? Made a ton of work easy for me as the heavy use of $ would have indicated :smile: Showing nothing when logged out was a confusing choice for me as well. What would you rather see there?

I'm working on the rest of the suggestions

jagannath93 commented on js/background.js:
```diff
@@ -1,80 +1,76 @@
-            chrome.tabs.reload(_tabs[j]);
-          }
-      });
-      ******************************************************************************************/
```
Try to figure out how to handle this TODO part. It would be awesome if you can implement this

dhruvkb replied:
I tried. Put in two days trying to make it work but the thing is Channel-i handles logout via AJAX, nothing changes on the client side and so there's no way the extension can know if the user has logged out.

jagannath93 replied:
Okay :smiley:

dhruvkb replied:
I managed to do this! If the tab logs out without AJAX, other pages are instantly refreshed to log them out as well and if the tabs logs out via AJAX other tabs are refreshed in the next minute as a part of the routine checkSession.

jagannath93 commented on js/popup.js:
```diff
@@ -1,228 +1,99 @@
-      var html_content = localStorage['apps_content'];
-      $('#main').html(html_content);
-    });
-  }
```
We've written this local storage part to fetch the recent app listing data from the backend irrespective of the state of the plugin on the client side for maintaining consistency. This is similar of versioning

So if you are fetching the app listing template & data dynamically from the backend then you have to implement this kind of versioning.

jagannath93 commented on css/popup.css:
```diff
@@ -1,177 +1,57 @@
+#img-logo-footer {
+    height: 30px;
+    width: 30px;
+}
```
Try to make this css cross-browser compatible (atleast for Google Chrome, Mozilla Firefox and Safari)

dhruvkb committed:
Move functions out of $(document).ready() and remove another definition of the identical function getHostName()

dhruvkb committed:
Replace 0/1 variables with boolean and make functions consistent by replacing var v = function() {} with function v() {}

ritz078 commented:
jQuery doesn't add any value to a developer's profile. JavaScript is easier and more widely used now. 

Show them the app links that can be visited even without logging in.

dhruvkb commented:
I'll take some time to remove all usages of jQuery and replace them with vanilla JavaScript. If not anything else, it's going to make the extension size smaller and my sphere of experience bigger :+1: :smile:

jagannath93 commented on popup.html:
```diff
@@ -1,51 +1,75 @@
+<script type="text/javascript" src="js/popup.js"></script>
+<script type="text/javascript" src="js/apps.js"></script>
+
 </html>
```
It's great that you've taken this up @dhruvkb :+1:

dhruvkb replied:
To be honest, it was greedy... I wanted to be an organization member getting my pull request merged was the only viable option :wink:

jagannath93 replied:
:smiley:

ritz078 replied:
btw I never became the member of the organisation 😄

dhruvkb replied:
@ritz078 I want to so I gotta do this

akki commented:
Is this PR going to resolve https://github.com/IMGIITRoorkee/Pocket-Channeli/issues/1?
The issue should also be update (closed) when this PR gets merged, if that's the case. Thanks!

dhruvkb commented:
This PR was going to resolve everything 😢 But it got delayed and now I'm getting beaten up by ETEs

dhruvkb committed:
Replace the array of global App objects with a dictionary of App objects and rework the loop to accommodate the dictionary instead of the array

akki commented:
Not a problem, resume work after the ETEs whenever you can. And all the best for them. :)

dhruvkb committed:
Eliminate all usages of jQuery, replacing it with vanilla JavaScript and also remove Semantic UI JS along with it

dhruvkb commented:
TODOs:
1. Add local storage, versioning and periodic update of apps via a list fetched from Channel-i
2. Show links to apps even when not logged in instead of a useless login page
3. Remove unnecessary Semantic UI CSS

Did I miss something?

dhruvkb committed:
Change NET_CHECK_DOMAIN to the media server running on NGINX and remove the use of info.cern.ch

dhruvkb committed:
Remove the login page and add the login functionality to the existing main page

dhruvkb committed:
Fix a counting error in the for loop running from 0 to length (both inclusive) by changing it to 0 (inclusive) to length (exclusive)

dhruvkb committed:
Fix a logical error in the handling of readystate by splitting the 'if DONE && 200' statement into 'if DONE' { if '200' } statements and keeping the 'else' with the inner 'if'

dhruvkb committed:
Set up a system to monitor login status across tabs and every minute and refresh tabs when the login status changes

rohithasrk commented:
What's the status of the app right now?

dhruvkb commented:
It's hard to fix and might need extensive APIs from Channel i to work, even after which people might not use it that much. We need more compelling features and a full-fledged API. I'm having second thoughts to taking out valuable time for something that might not help more than a few people.

dhruvkb commented:
Found a JSON endpoint in Channel i that returns a list of all applications and their URLs (also games, links and IMG tools as a bonus). This extension is back in the works.

akki commented:
@dhruvkb Make sure to notify the maintainers of that JSON endpoint that you will be using it for `Pocket-Channeli`.

dhruvkb committed:
Shift from hardcoded list of apps to Internet loaded list of apps, updated everytime the status of being signed in changes

dhruvkb committed:
Major changes throughout the entire project
Add tabs for games and IMG tools
Increase width for more appealing UI
Remove user card and other unnecessary UI elements
Split large files into smaller, more manageable chunks

dhruvkb committed:
isUserLoggedIn becomes userIsLoggedIn in line with boolean variable naming conventions

dhruvkb committed:
Add a button to log in or log out based on the value of userIsLoggedIn

dhruvkb committed:
Modify screenshots to reflect the modified view

dhruvkb committed:
Fix a typo in the screenshot links

dhruvkb commented:
@akki @ritz078 @jagannath93 It works great now, uses Chrome's local storage, updates app list over the Internet and synchronises the user's logged in status across all open Channeli pages (within 1 minute).

@rohithasrk I found an endpoint on Channel i (/get_links/) that does exactly what I needed so I take back my word that this extension is dead.
